### PR TITLE
[base] Fixed possible deadlock on WorkerThread shutdown.

### PR DESCRIPTION
**Pull request by ygorshenin**

### CODE REVIEW

File: base/base_tests/worker_thread_tests.cpp [hunk 1]
```diff
@@ -67,4 +68,22 @@ UNIT_TEST(WorkerThread_SimpleFlush)
...
+
+UNIT_TEST(WorkerThread_PushFromPendingTask)
+{
+  // promise - future pair is used as a socketpair here to pass a
+  // signal from the main thread to the worker thread.
+  promise<void> p;
+  auto f = p.get_future();
+
+  WorkerThread thread;
+  bool const rv = thread.Push([&]() {
```
Comment (mgsergio) [inline base/base_tests/worker_thread_tests.cpp]: [&thread, f]

File: base/base_tests/worker_thread_tests.cpp [hunk 2]
```diff
@@ -67,4 +68,22 @@ UNIT_TEST(WorkerThread_SimpleFlush)
...
+    f.get();
+    bool const rv = thread.Push([&]() { TEST(false, ("This task should not be executed")); });
```
Comment (mgsergio) [inline base/base_tests/worker_thread_tests.cpp]: []

Comment (ygorshenin) [inline base/base_tests/worker_thread_tests.cpp]: Done

File: base/base_tests/worker_thread_tests.cpp (hunk 1 above)
Comment (ygorshenin) [inline base/base_tests/worker_thread_tests.cpp]: Okay

Comment (ygorshenin) [inline base/base_tests/worker_thread_tests.cpp]: From my point of view, it's not a problem to capture f by reference.

**Comment by ygorshenin** [comment]
PTAL

**Pull request merged**